syntax = "proto3";

package coprocess;

option go_package = "/coprocess";

 
// Plugin hooks enumeration that identifies the type of plugin
enum HookType {
  
  // Use for error checking and handling of an unrecognised hook type.
  Unknown = 0;
  
  // Executed before request sent to upstream target and before any
  // authentication information is extracted from the header or
  // parameter list of the request. Applies to both keyless and protected
  // APIs.
  Pre = 1;

  // Executed after authentication, validation, throttling and quota-limiting
  // middleware has been executed, just before the request is proxied upstream. Use this
  // to post-process a request before sending it to upstream API. This is only called
  // when using protected APIs.
	Post = 2;

  // Executed after authentication, validation, throttling, and quota-limiting
  // middleware has been executed, just before the request is proxied upstream. Use this
  // to post-process a request before sending it to upstream API. This is only called
  // when using protected APIs.
	PostKeyAuth = 3;
	
  // Executed for performing customised authentication.
  CustomKeyCheck  = 4;

  // Executed after the upstream API replies. The arguments passed to this hook include
  // both the request and response data. Use this to modify the HTTP response before itâ€™s
  // sent to the client. This hook also receives the request object, the session object,
  // the metadata and API definition associated with the request.
  Response = 5;
}

// List of strings
message StringSlice {

  // List of string items
  repeated string items = 1;
}
